### PR TITLE
Generic stochastic model

### PR DESCRIPTION
**Pull request by phumtutum**
In this pr:
- A generic stochastic model is added (MarkovJumpModel) which uses Gillespie's algorithm
- An example of a model that illustrates how simply other stochastic models can be added from now on (Michaelis Menten model)
- Added a degradation model for testing as well
- Removed old degradation model

### CODE REVIEW

**Comment by phumtutum** [comment]
Now, the old degradation model is removed and the pr is ready for review

File: docs/source/toy/stochastic/index.rst [hunk 1]
```diff
@@ -0,0 +1,15 @@
+***********************
+Stochastic Toy Problems
+***********************
+
+The `stochastic` module provides toy :class:`models<pints.toy.stochastic.MarkovJumpModel>`,
```
Comment (MichaelClerx) [inline docs/source/toy/stochastic/index.rst:5]: I guess initially it's just models?

File: examples/README.md [hunk 2]
```diff
@@ -117,6 +117,7 @@ relevant code.
 - [SIR Epidemiology model](./toy/model-sir.ipynb)
 - [Stochastic Degradation model](./toy/model-stochastic-degradation.ipynb)
 - [Stochastic Logistic model](./toy/model-stochastic-logistic-growth.ipynb)
+- [Michaelis Menten model](./toy/model-stochastic-michaelis-menten.ipynb)
```
Comment (MichaelClerx) [inline examples/README.md:120]: Separate category here too?

**Comment by phumtutum** [comment]
This pr updates the stochastic degradation model. This model is used in a ipynb in https://github.com/pints-team/pints/pull/1413 and therefore changes need to be made to that ipynb in this pr once that pr is merged into master. I will make the requested changes at once with those changes.

File: pints/toy/stochastic/_markov_jump_model.py [hunk 3]
```diff
@@ -0,0 +1,149 @@
+#
+# Stochastic degradation toy model.
```
Comment (MichaelClerx) [inline pints/toy/stochastic/_markov_jump_model.py:2]:
```suggestion
# Markov jump model.
```

File: pints/toy/stochastic/_michaelis_menten_model.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
+#
+# Stochastic degradation toy model.
```
Comment (MichaelClerx) [inline pints/toy/stochastic/_michaelis_menten_model.py:2]:
```suggestion
# Stochastic michaelis-menten toy model.
```

**Review by MichaelClerx — APPROVED** [review]
Fine with me, if @ben18785 likes it!

File: examples/README.md (hunk 2 above)
Comment (ben18785) [reply examples/README.md:120]: Yes, I think separate out the stochastic ones into their own subsection of "toy models".

File: pints/toy/stochastic/_degradation_model.py [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+        k = parameters[0]
+
+        if k <= 0:
+            raise ValueError('Rate constant must be positive.')
+
+        times = np.asarray(times)
+        if np.any(times < 0):
+            raise ValueError('Negative times are not allowed.')
+
+        variance = np.exp(-2 * k * times) * (-1 + np.exp(k * times)) * self._x0
```
Comment (ben18785) [inline pints/toy/stochastic/_degradation_model.py:78]: I checked -- this is correct

File: pints/toy/stochastic/_degradation_model.py [hunk 6]
```diff
@@ -0,0 +1,87 @@
...
+        k = parameters[0]
+
+        if k <= 0:
+            raise ValueError('Rate constant must be positive.')
+
+        times = np.asarray(times)
+        if np.any(times < 0):
+            raise ValueError('Negative times are not allowed.')
+
+        mean = self._x0 * np.exp(-k * times)
```
Comment (ben18785) [inline pints/toy/stochastic/_degradation_model.py:58]: Correct

File: pints/toy/stochastic/_markov_jump_model.py [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+import random
+
+from .. import ToyModel
+
+
+class MarkovJumpModel(pints.ForwardModel, ToyModel):
+    r"""
+    A general purpose Markov Jump model used for any systems of reactions
+    that proceed through jumps. We simulate a population of N different species
+    reacting through M different mechanisms
```
Comment (ben18785) [inline pints/toy/stochastic/_markov_jump_model.py:20]: "reacting through M different reaction equations"

File: pints/toy/stochastic/_markov_jump_model.py [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+
+    1. Sample values :math:`r_0`, :math:`r_1`, from a uniform distribution
+
+    .. math::
+        r_0, r_1 \sim U(0,1)
+
+    2. Calculate the time :math:`\tau` until the next single reaction as
+
+    .. math::
+        \tau = \frac{-\ln(r)}{a_0}
```
Comment (ben18785) [inline pints/toy/stochastic/_markov_jump_model.py:32]: We haven't said what a_0 is.

File: pints/toy/stochastic/_markov_jump_model.py [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+
+    3. Decide which reaction, i, takes place using r_1 * a_0 and iterating
+
+    through propensities
```
Comment (ben18785) [inline pints/toy/stochastic/_markov_jump_model.py:36]: This needs quite a bit more explanation, especially since this is the key point of the algorithm

**Review by ben18785 — CHANGES_REQUESTED** [review]
Thanks @phumtutum -- looks really good. There are a few things that I think would be good to change, then I'm happy to merge.

In the degradation notebook:

- I think this line needs rewriting: "averaging the concentration values at each time step, produces a reproducible result" -- I don't think that "reproducible result" is the correct terminology here. I guess you just want to say that the mean of the distribution is that exponential thingy?

- "The deterministic mean (from above) is plotted with the deterministic standard deviation" -- I think this is a bit ambiguous. I think you want to say, "We now plot the analytic mean and standard deviation of this process."

In the michaelis-menten notebook:

- "The Michaelis Menten model is a stochastic model, more details can be found here.": the MM model actually isn't necessarily a stochastic model (we have a deterministic version of a similar but not same version of the model in the package), and the stuff you link to is about the deterministic model. So, can we say, "This is a stochastic version of the Michaelis-Menten model, and more details about the deterministic analogue can be found here."
- Can you also provide a little bit of a description about what this model represents in the first part of this notebook?
- "This function, given a set of parameters and times, computes the appropriate times and values using Gillispie's algorithm and then uses interpolation to find the values at the exact times requested" -- it's not clear what the values are (i.e. they are counts of molecules) and "Gillespie" has a typo. I think you also need to explain what the interpolation does and why it is needed. I've noticed "Gillespie" is misspelled throughout the notebook actually.
- "Given the stochastic nature of this model we can use multiple simulations to make sure that the runs are covering the same model with the same parameters. Our simulations are close, suggesting we are obtaining accurate simulations." -- this isn't correct, really. All you are showing is that there is not a large amount of stochasticity in these models for these initial values. Indeed, can we show that when we use much smaller molecule counts as initial conditions that we get much more stochasticity?
- It would be good to show that the solutions with high molecular counts correspond with the solutions of the corresponding ODE: you could code this up using scipy's `odeint` within the notebook

**Comment by phumtutum** [comment]
@ben18785 I have addressed your comments. If there is anything else I should change please let me know.

**Review by ben18785 — APPROVED** [review]

**Pull request merged**